File: src/engraft/actions/html_replace.py
import re
from dataclasses import dataclass, field
from pathlib import Path

from lxml import html as lxml_html
from lxml.etree import _Element

from engraft.actions import register
from engraft.actions.base import Action


def _extract_doctype(raw_html: str) -> str | None:
    """Extract the DOCTYPE declaration from raw HTML, if present."""
    match = re.match(r"(<!DOCTYPE[^>]*>)", raw_html, re.IGNORECASE)
    return match.group(1) if match else None


@register("html_replace")
@dataclass
class HtmlReplace(Action):
    """Replace values in HTML files using XPath selectors."""

    file: str
    replace: list[dict[str, str]] = field(default_factory=list)

    def apply(
        self,
        variables: dict[str, str],
        work_dir: Path,
        values_dir: Path,
    ) -> None:
        target = work_dir / self.file
        raw_html = target.read_text()
        doctype = _extract_doctype(raw_html)

        doc = lxml_html.fromstring(raw_html)

        for entry in self.replace:
            selector = entry["selector"]
            var_name = entry["variable"]
            value = variables[var_name]

            results = doc.xpath(selector)

            if len(results) == 0:
                raise ValueError(
                    f"XPath selector {selector!r} matched no elements in {self.file}"
                )
            if len(results) > 1:
                raise ValueError(
                    f"XPath selector {selector!r} matched {len(results)} elements "
                    f"in {self.file}, expected exactly 1"
                )

            result = results[0]

            if isinstance(result, _Element):
                # Element node — set text content
                result.text = value
            else:
                # Attribute result — lxml returns _ElementStringResult
                # which has getparent() and attrname
                parent = result.getparent()
                attr_name = result.attrname
                parent.set(attr_name, value)

        # Serialize back to HTML
        output = lxml_html.tostring(doc, encoding="unicode", method="html")

        if doctype:
            output = doctype + "\n" + output + "\n"
        else:
            output = output + "\n"

        target.write_text(output)

    def target_files(self) -> list[str]:
        """Return project-relative file paths this action operates on."""
        return [self.file]

File: typescript/src/actions/htmlReplace.ts
import { promises as fs } from "node:fs";
import path from "node:path";

import { JSDOM } from "jsdom";

import type { Action, Variables } from "./base.js";
import { register } from "./registry.js";

interface ReplaceEntry {
  selector: string;
  variable: string;
}

const DOCTYPE_RE = /^\s*<!DOCTYPE[^>]*>/i;

const extractDoctype = (html: string): string | null => {
  const match = DOCTYPE_RE.exec(html);
  return match ? match[0].replace(/^\s+/, "") : null;
};

const stripDoctype = (html: string): string => html.replace(DOCTYPE_RE, "");

@register("html_replace")
export class HtmlReplace implements Action {
  private readonly file: string;
  private readonly replace: ReplaceEntry[];

  constructor(config: Record<string, unknown>) {
    this.file = String(config.file);
    this.replace = (config.replace as ReplaceEntry[] | undefined) ?? [];
  }

  async apply(
    variables: Variables,
    workDir: string,
    _valuesDir: string,
  ): Promise<void> {
    const target = path.join(workDir, this.file);
    const rawHtml = await fs.readFile(target, "utf8");
    const doctype = extractDoctype(rawHtml);
    const withoutDoctype = stripDoctype(rawHtml);

    const dom = new JSDOM(withoutDoctype, { contentType: "text/html" });
    const { document } = dom.window;
    const XPathResult = dom.window.XPathResult;

    for (const entry of this.replace) {
      const value = variables[entry.variable];
      if (value == null) continue;

      const result = document.evaluate(
        entry.selector,
        document,
        null,
        XPathResult.ORDERED_NODE_SNAPSHOT_TYPE,
        null,
      );

      const count = result.snapshotLength;
      if (count === 0) {
        throw new Error(
          `XPath selector ${JSON.stringify(entry.selector)} matched no elements in ${this.file}`,
        );
      }
      if (count > 1) {
        throw new Error(
          `XPath selector ${JSON.stringify(entry.selector)} matched ${count} elements in ${this.file}, expected exactly 1`,
        );
      }

      const node = result.snapshotItem(0)!;
      if (node.nodeType === dom.window.Node.ATTRIBUTE_NODE) {
        (node as Attr).value = value;
      } else if (node.nodeType === dom.window.Node.ELEMENT_NODE) {
        (node as Element).textContent = value;
      } else {
        throw new Error(
          `XPath selector ${JSON.stringify(entry.selector)} resolved to an unsupported node type`,
        );
      }
    }

    const body = dom.serialize();
    const output = doctype ? `${doctype}\n${body}\n` : `${body}\n`;
    await fs.writeFile(target, output);
  }

  targetFiles(): string[] {
    return [this.file];
  }
}
